### ISSUE TITLE
Sub-millisecond request latency metrics

### ISSUE THREAD

**Issue by seizethedave** [envoyproxy/envoy#21429]
*Title*: Sub-millisecond request latency metrics

*Description*:
My upstream service has a typical response time of 1-2ms, so having all of the request latency statistics (like `upstream_rq_time`) truncated to nearest whole millisecond is not ideal. It would be great if I could have these reported in microseconds/etc.

Thanks!

**Comment by htuch** [comment]
Seems reasonable to introduce some sub-millisecond precision. CC @jmarantz @mattklein123

**Comment by mattklein123** [comment]
I think this is a fine request. The only thing I would point out is depending on how it is expressed, there might be some issues with the default Prometheus histogram buckets that need sorting through. cc @ggreenway

**Comment by ggreenway** [comment]
I think the default buckets will end up being fine, and the buckets are defined as doubles so fractional values can be configured.

The trickier part is that many of our time-based metrics are defined with a base-unit of milliseconds, and the histogram datatype is an integer, so there's no way with the current stats to record a value of less than 1 millisecond. So we'd need to replace those metrics with a microsecond-version of them, or switch the internal representation to be a double instead of uint64.

**Comment by mattklein123** [comment]
Could we just cheat and store microseconds instead of millis and then convert or just have a different metric?

**Comment by ggreenway** [comment]
Yes, I think so. I wrote some code to do scaled Percent in histograms. We could add something similar for this. See: 
```
$ git grep PercentScale
envoy/stats/histogram.h:             // PercentScale to get the actual value, eg a value of 100% (or 1.0) is encoded as
envoy/stats/histogram.h:             // PercentScale, 50% is encoded as PercentScale * 0.5. Encoding as fixed-point allows
envoy/stats/histogram.h:  static constexpr uint64_t PercentScale = 1000000;
source/common/stats/histogram_impl.cc:  constexpr double percent_scale = Histogram::PercentScale;
source/extensions/stat_sinks/common/statsd/statsd.cc:    constexpr float divisor = Stats::Histogram::PercentScale;
source/extensions/stat_sinks/common/statsd/statsd.cc:    constexpr float divisor = Stats::Histogram::PercentScale;
source/extensions/transport_sockets/tcp_stats/tcp_stats.cc:                  (Stats::Histogram::PercentScale < UINT32_MAX));
source/extensions/transport_sockets/tcp_stats/tcp_stats.cc:         static_cast<uint64_t>(Stats::Histogram::PercentScale)) /
```

**Comment by ashishb-solo** [comment]
I'm currently making an attempt at this using the strategy outlined above, namely storing everything in microseconds, but providing the option to display in milliseconds (as default to preserve backward compatibility). I'm finding it fairly straightforward to modify how data is stored in the histogram, however it's not immediately obvious to me how I would display the histogram in milliseconds if the data are stored in microseconds. Could I have some pointers on how I might achieve this? Thanks!

**Comment by ggreenway** [comment]
Can you post a PR with what you have so far? That will make it easier to understand what specifically you need help with.
